docker/incremental_load.sh.tpl

```diff
@@ -22,38 +22,96 @@ RUNFILES="${PYTHON_RUNFILES:-${BASH_SOURCE[0]}.runfiles}"
 
 DOCKER="${DOCKER:-docker}"
 
-# List all images identifier (only the identifier) from the local
-# docker registry.
-IMAGES="$("${DOCKER}" images -aq)"
+# Fetch the diff ids of the layers loaded in the docker daemon already
+IMAGES=$(${DOCKER} inspect $(${DOCKER} images -aq) |& \
+  grep sha256 | grep -v Id | cut -d'"' -f 2 | cut -d':' -f 2 | \
+  sort | uniq)
 IMAGE_LEN=$(for i in $IMAGES; do echo -n $i | wc -c; done | sort -g | head -1 | xargs)
 
 [ -n "$IMAGE_LEN" ] || IMAGE_LEN=64
 
-function incr_load() {
+# Create temporary files in which to record things to clean up.
+TEMP_FILES="$(mktemp)"
+TEMP_IMAGES="$(mktemp)"
+function cleanup() {
+  cat "${TEMP_FILES}" | xargs rm -rf> /dev/null 2>&1 || true
+  cat "${TEMP_IMAGES}" | xargs "${DOCKER}" rmi > /dev/null 2>&1 || true
+
+  rm -rf "${TEMP_FILES}"
+  rm -rf "${TEMP_IMAGES}"
+}
+trap cleanup EXIT
+
+
+function load_legacy() {
+  local tarball="${RUNFILES}/$1"
+
+  # docker load has elision of preloaded layers built in.
+  echo "Loading legacy tarball base $1..."
+  "${DOCKER}" load -i "${tarball}"
+}
+
+function import_layer() {
   # Load a layer if and only if the layer is not in "$IMAGES", that is
   # in the local docker registry.
-  name=$(cat ${RUNFILES}/$1)
+  local name="$(cat "${RUNFILES}/$1")"
 
   if (echo "$IMAGES" | grep -q ^${name:0:$IMAGE_LEN}$); then
-    echo "Skipping $name, already loaded."
+    echo "Skipping ${name}, already loaded."
   else
-    echo "Loading $name..."
-    "${DOCKER}" load -i ${RUNFILES}/$2
+    echo "Importing ${name}..."
+    TEMP_IMAGE="$(cat "${RUNFILES}/$2" | "${DOCKER}" import -)"
+    echo "${TEMP_IMAGE}" >> "${TEMP_IMAGES}"
   fi
 }
 
+function import_config() {
+  # Create an image from the image configuration file.
+  local name="${RUNFILES}/$1"
+  local diff_id="$(cat "${RUNFILES}/$2")"
+  local layer="${RUNFILES}/$3"
+
+  local tmp_dir="$(mktemp -d)"
+  echo "${tmp_dir}" >> "${TEMP_FILES}"
+
+  cd "${tmp_dir}"
+  cp "${name}" the-image.json
+  cp "${layer}" "${diff_id}.tar"
+  local parent_layers=$(cat the-image.json | python -mjson.tool | \
+    grep sha256 | grep "," | cut -d'"' -f 2 | cut -d':' -f 2)
+
+  cat > manifest.json <<EOF
+[{
+   "Config": "the-image.json",
+   "Layers": [$(
+for x in ${parent_layers}
+do
+  # We must follow a consistent scheme here in case it overlaps
+  # with the final layer.
+  echo -n "\"$x.tar\","
+done
+echo -n "\"${diff_id}.tar\""
+)
+   ],
+   "RepoTags": []
+}]
+EOF
+  tar cf image.tar the-image.json "${diff_id}.tar" manifest.json
```

Comment:
Any reason to write image.tar to disk instead of piping it right into docker load?